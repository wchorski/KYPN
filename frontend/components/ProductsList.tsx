import Image from 'next/image';
import Link from 'next/link';
import { ProductThumbnail } from "@/components/ProductThumbnail";
import { useQuery, gql } from '@apollo/client'
import styled from 'styled-components';
import { QueryLoading } from './menus/QueryLoading';
import { QueryError } from './menus/QueryError';

export function ProductsList() {
  const { loading, error, data } = useQuery(GET_ALL_PRODUCTS);

  if (loading) return <QueryLoading />
  if (error) return <QueryError />

  return (
    <StyledProductsList>
      {data.products.map((prod: any) => {
        return (
          <li key={prod.id}>
            <ProductThumbnail {...prod} />

          </li>
        );
      })}
    </StyledProductsList>
  )
}

const StyledProductsList = styled.ul`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
  list-style: none;
  margin: 0;
  padding: 0;

  li{
    background-color: #dcdcdc;
    padding: .3em;
  }

  img{
    width: 100%;
    height: auto;
  }
`

export const GET_ALL_PRODUCTS = gql`
  query Products {
    products {
      id
      description
      name
      photo {
        image {
          url
          width
          height
          id
        }
        id
        altText
      }
      price
      status
    }
  }
`