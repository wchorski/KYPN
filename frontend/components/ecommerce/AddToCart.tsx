'use client'
import { useMutation, gql } from '@apollo/client';
import { useRouter } from 'next/navigation';
import { useSession } from "../menus/Session";
import { MdShoppingBag } from 'react-icons/md';
import { TbCheck, TbExclamationCircle, TbQuestionMark, TbLoader  } from 'react-icons/tb';
import { QUERY_USER_CURRENT } from '../menus/Session';
import { useState } from 'react';
import styles from '../../styles/eyecandy/SpinCycle.module.scss'
import stylesButton from "@/styles/ecommerce/AddToCart.module.scss";

const delay = (ms:number) => new Promise(res => setTimeout(res, ms));

type State = 'loading'|'pending'|'error'|'out_of_stock'|'success'|undefined

export default function AddToCart({ id }: { id: string }) {

  const [state, setstate] = useState<State>(undefined)
  const session = useSession()
  const router = useRouter()
  const [addToCart, { loading }] = useMutation(ADD_TO_CART_MUTATION)


  async function handleButton() {

    if (!session) return router.push(`/auth/login`)

    try {

      setstate('pending')
      
      const res = await addToCart({
        variables: {
          addToCartId: id,
          productId: id
        },
        refetchQueries: [{ query: QUERY_USER_CURRENT }],
      })

      await delay(500)

      setstate('success')

      await delay(500)

      setstate(undefined)
      
    } catch (error) {
      setstate('error')
      console.warn('addtocart error: ', error);
    }


    // console.log({res});

  }

  const renderIcon = (state:State) => {

    switch (state) {
      case 'pending':
        return <TbLoader className={styles.spin}/>
        
      case 'success':
        return <TbCheck />

      case 'error':
        return <TbExclamationCircle />
    
      default:
        return <MdShoppingBag /> 
    }
  }

  return (<>
    <button 
      type="button" 
      disabled={loading || state !== undefined} 
      onClick={e => handleButton()}
      className={' addtocart' + ' button'}
    >
      <span>Add To Cart <span style={{marginRight: 'auto'}} className='state'>{renderIcon(state)}</span> </span>
      
    </button>
    
    </>);
}

const ADD_TO_CART_MUTATION = gql`
  mutation Mutation($addToCartId: ID!, $productId: ID) {
    addToCart(id: $addToCartId, productID: $productId) {
      id
      quantity
      product {
        name
      }
    }
  }
`
