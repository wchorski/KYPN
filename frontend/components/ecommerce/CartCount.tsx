import styled from "styled-components"
import {CSSTransition, TransitionGroup,} from 'react-transition-group';
import { useRef } from "react";
import { MdShoppingBag } from "react-icons/md";

export default function CartCount({count}:{count:number}) {

  const nodeRef = useRef(null);

  return (
    <StyledAnimation>
      <TransitionGroup>
        <CSSTransition
          nodeRef={nodeRef}
          className={'count'}
          classNames={'count'}
          key={count}
          timeout={{enter: 400, exit: 400}}
        >
          <StyledDot ref={nodeRef}>
            {count}
            <MdShoppingBag />
          </StyledDot>
        </CSSTransition>
      </TransitionGroup>
    </StyledAnimation>
  )
}

const StyledDot = styled.div`
  background-color: var(--c-1);
  color: var(--c-txt-rev);
  padding: 0.5rem;
  border-radius: 50px;
  line-height: 2rem;
  min-width: 3rem;
  margin-left: 1rem;
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;

`
// TODO this animation is wack
const StyledAnimation = styled.span`
  position: relative;

  .count{
    display: block;
    position: relative;
    transition:  0.4s;
    backface-visibility: hidden;
    transition: 0.4s;
    /* position: absolute; */
    top: 0;
    left: 0;
  }

  .count-enter{
    /* background-color: green; */
    opacity: 0;
    transform: scale(1.3) rotateX(45deg);
  }

  .count-enter-active {
    /* background-color: red; */
    opacity: 1;
    /* transition: 2s; */
    /* transform: rotateX(0); */
  }

  .count-exit{
    /* background-color: blue; */
    opacity: 0;
    top: 0;
    position: absolute;
    /* transform: rotateX(0); */
  }

  .count-exit-active{
    /* background-color: pink; */
    opacity: 1;
    /* transition: 2s; */
    transform: scale(1) rotateX(0);
  }
`