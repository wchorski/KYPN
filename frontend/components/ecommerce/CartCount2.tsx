import styled from "styled-components"
import { MdShoppingBag } from "react-icons/md";


export function CartCount2({ count }: { count: number }) {

  return (
    <StyledCartIcon className="cartCount">

      <StyledDot key={count}>
        <span> {count} </span>
      </StyledDot>

      <MdShoppingBag />

    </StyledCartIcon>
  )
}


const StyledDot = styled.div`

  background-color: var(--c-txt-rev);
  color: var(--c-txt);
  padding: 0.2rem;
  position: absolute;
  top: -10px;
  right: 0;
  border-radius: 100px;
  line-height: 2rem;
  min-width: 2.5rem;
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
  animation: bigSmall .3s;
  animation-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  outline: solid white 1px;


  @keyframes bigSmall {
    0% { 
      font-size: 25px;
      padding: 1rem;
    }
    100% { 
      font-size: 2px;
      padding: 0.2rem;
    }
  }

  span {

    transition: all .3s;
  }
`

const StyledCartIcon = styled.div`
  position: relative;
`