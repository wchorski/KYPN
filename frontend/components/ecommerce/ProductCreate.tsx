// import { uploadPrep } from "@/lib/uploadPrep"
import Router from "next/router";
import useForm from "@/lib/useForm"
import { StyledForm } from "@/styles/Form.styled"
import { gql, useMutation } from "@apollo/client"
import { useState } from "react"
import ErrorMessage from "../ErrorMessage"
import { GET_ALL_PRODUCTS } from "../ProductsList"


export const ProductCreate = () => {

  // const [mutate] = useMutation(FILE_UPLOAD_MUTATION, {
  //   refetchQueries: [{ query: filesQuery}]
  // })

  // function handleOnChange(e: any){
  //   console.log('upload change');
  //   console.log(e.target.files[0]);
    
  // } 

  const [isSuccess, setIsSuccess] = useState(false)

  const {inputs, handleChange, clearForm, resetForm} = useForm({
    image: null,
    name: 'will',
    price: 123,
    description: "mydesc"
  })

  async function handleSubmit(e: any) {
    e.preventDefault()
    console.log({inputs})
    const res = await createProduct()
    console.log('res', res)
    if(res.data.createProduct) clearForm(); setIsSuccess(true)
    Router.push({
      pathname: `/shop/product/${res.data.createProduct.id}`,
    })    
  }

  
  const [createProduct, {loading, error, data}] = useMutation(
    CREATE_PRODUCT_MUTATION, 
    {
      variables: {
        data: {
          photo: {
            create: {
              altText: `${inputs.name} featured image`,
              image: {upload: inputs.image}
            },
          },
          name: inputs.name,
          price: inputs.price,
          description: inputs.description,
        }
      },
      // ? after new product is added, fetch re runs in background so client doesn't half to hard refresh
      refetchQueries: [{query: GET_ALL_PRODUCTS}]
    }
  )


  return (<>
    {/* <input type="file" multiple required onChange={e => handleOnChange(e)} /> */}

    <StyledForm onSubmit={e => handleSubmit(e)}>

      <ErrorMessage error={error}/>
      {isSuccess && <p>product created</p> }

      <fieldset disabled={loading} aria-busy={loading} >
        <label htmlFor="image">
          Image
          <input type="file" id="image" name="image" 
            onChange={handleChange}
          />
        </label>

        <label htmlFor="name">
          Name
          <input required type="text" id="name" name="name" placeholder="name..." 
            value={inputs.name}
            onChange={handleChange}
          />
        </label>

        <label htmlFor="price">
          Price
          <input required type="number" id="price" name="price" placeholder="price..." 
            value={inputs.price}
            onChange={handleChange}
          />
        </label>

        <label htmlFor="description">
          Description
          <textarea id="description" name="description" placeholder="description..." 
            value={inputs.description}
            onChange={handleChange}
          />
        </label>

        <button type="submit" > + Add </button> 
        <br /> <br />

        <button type="button" onClick={e => clearForm()}> clear form </button>
        {/* <button type="button" onClick={e => resetForm()}> reset form </button> */}

      </fieldset>
    </StyledForm>
  </>)
}

// const FILE_UPLOAD_MUTATION = gql`
//   mutation ($file: Upload!) {
//     uploadFile(file: $file) {
//       success
//     }
//   }
// `;

const CREATE_PRODUCT_MUTATION = gql`
  mutation CreateProduct($data: ProductCreateInput!) {
    createProduct(data: $data) {
      name
      photo {
        altText
        id
        image {
          extension
          filesize
          height
          id
          url
          width
        }
      }
      id
      description
      price
      status
    }
  }
`
