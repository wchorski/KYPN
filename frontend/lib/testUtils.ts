import casual from 'casual';
import { QUERY_PRODUCTS_COUNT } from '../components/Pagination';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  // __typename: 'Item',
  id: 'abc123',
  price: 5000,
  user: null,
  photo: {
    id: 'abc123',
    altText: 'dogs are best',
    image: {
      url: 'dog.jpg',
    },
  },
  name: 'dogs are best',
  description: 'dogs',
});

const fakeUser = (overrides: any | void) => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
  ...overrides,
});

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2022-12-11T20:16:13.797Z',
  // @ts-ignore
  user: fakeUser(),
});

const fakeCartItem = (overrides: any | void) => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  product: fakeItem(),
  // @ts-ignore
  user: fakeUser(),
  ...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    // @ts-ignore
    this.store = {};
  }

  clear() {
    // @ts-ignore
    this.store = {};
  }

  getItem(key: any) {
    // @ts-ignore
    return this.store[key] || null;
  }

  setItem(key: any, value: any) {
    // @ts-ignore
    this.store[key] = value.toString();
  }

  removeItem(key: any) {
    // @ts-ignore
    delete this.store[key];
  }
}

function makePaginationMocksFor(length: any) {
  return [
    {
      request: { query: QUERY_PRODUCTS_COUNT },
      result: {
        data: {
          _allProductsMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              count: length,
              __typename: 'count',
            },
          },
        },
      },
    },
  ];
}

export {
  makePaginationMocksFor,
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
};
