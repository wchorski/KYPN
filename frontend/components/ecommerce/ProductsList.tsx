import { ProductThumbnail } from "@components/ecommerce/ProductThumbnail";
import { gql } from '@apollo/client'
import { QueryLoading } from '@components/menus/QueryLoading';
import { QueryError } from '@components/menus/QueryError';
import styles from '@/styles/ecommerce/Product.module.scss'
import { envvars } from "@lib/envvars";
import { getClient } from "@lib/gqlClient";
import { Product } from "@lib/types";
import { List } from "@components/elements/List";


type ProdProps = {
  page: number,
  categoryNames:string[],
  products:Product[],
}

export async function ProductsList({ products = [] }: ProdProps) {


  // console.log({ data });

  return (
    <List isAnimated={true} className={styles.product}>
      {products?.map((prod: any) =>  (
        <ProductThumbnail {...prod} />
      ))}
    </List>
  )
}