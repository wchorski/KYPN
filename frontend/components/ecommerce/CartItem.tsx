import moneyFormatter from "../../lib/moneyFormatter"
import { StyledCartItem } from "../../styles/CartItem.styled"
import Image from "next/image"
import CartRemoveItem from "./CartRemoveItem"
import { handlePhoto } from "../../lib/handleProductPhoto"
import { ImageDynamic } from "../elements/ImageDynamic"

export default function CartItem({ item }: any) {


  if (!item?.product) return (
    <StyledCartItem>
      <p>This cart item is no longer supplied by our store</p>
    </StyledCartItem>
  )


  const { product: { id, description, name, price, photo, }, quantity } = item


  return (
    <StyledCartItem>
      {/* <Image
        priority
        src={handlePhoto(photo).image?.url}
        alt={handlePhoto(item.photo).image?.altText ? handlePhoto(item.photo).image?.altText : 'no product photo'}
        width={handlePhoto(photo).image?.width}
        height={handlePhoto(photo).image?.height}
      /> */}

      <ImageDynamic photoIn={photo} />

      <h5>{name}</h5>

      <div className="perItemTotal">
        <p>{moneyFormatter(price * quantity)}</p>

        <em>{quantity} &times; {moneyFormatter(price)} each </em>
      </div>

      <CartRemoveItem id={item.id} />
    </StyledCartItem>
  )
}