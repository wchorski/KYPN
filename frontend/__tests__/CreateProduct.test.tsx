import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MockedProvider } from "@apollo/react-testing";
import { withRouter } from "next-router-mock";
import { fakeUser, fakeItem, makePaginationMocksFor } from '../lib/testUtils'
import { ProductCreate, CREATE_PRODUCT_MUTATION } from "../components/ecommerce/ProductCreate";
import { GET_ALL_PRODUCTS } from "../components/ProductsList";

const newItem = fakeItem()

describe('<ProductCreate />', () => {

  it('render and match snapshot', () => {
    const { container, debug } = render(
      <MockedProvider >
        <ProductCreate />
      </MockedProvider>
    )

    expect(container).toMatchSnapshot()
    // debug()

  })

  it('handles and updating', async () => {
    const { container, debug } = render(
      <MockedProvider >
        <ProductCreate />
      </MockedProvider>
    )

    await userEvent.type(screen.getByPlaceholderText(/Name/i), newItem.name)
    // @ts-ignore
    await userEvent.type(screen.getByPlaceholderText(/Price/i), String(newItem.price))
    await userEvent.type(screen.getByPlaceholderText(/Description/i), newItem.description)

    expect(screen.getByDisplayValue(newItem.name)).toBeInTheDocument()
  });
})