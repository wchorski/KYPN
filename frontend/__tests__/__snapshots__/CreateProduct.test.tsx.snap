// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCreate /> render and match snapshot 1`] = `
<div>
  <form
    class="sc-dc2830a3-0 iEvsZN"
  >
    <fieldset
      aria-busy="false"
    >
      <label
        for="image"
      >
        Image
        <input
          id="image"
          name="image"
          type="file"
        />
      </label>
      <label
        for="name"
      >
        Name
        <input
          id="name"
          name="name"
          placeholder="name..."
          required=""
          type="text"
          value="will"
        />
      </label>
      <label
        for="price"
      >
        Price
        <input
          id="price"
          name="price"
          placeholder="price..."
          required=""
          type="number"
          value="123"
        />
      </label>
      <label
        for="description"
      >
        Description
        <textarea
          id="description"
          name="description"
          placeholder="description..."
        >
          mydesc
        </textarea>
      </label>
      <button
        type="submit"
      >
         + Add 
      </button>
      <br />
       
      <br />
      <button
        type="button"
      >
         clear form 
      </button>
    </fieldset>
  </form>
</div>
`;
